pkg/plugins/policies/meshtimeout/plugin/v1alpha1/plugin.go

```diff
@@ -263,48 +271,79 @@ func createInboundClusterName(servicePort uint32, listenerPort uint32) string {
 	}
 }
 
-func applyToRealResources(rs *core_xds.ResourceSet, rules outbound.ResourceRules, meshCtx xds_context.MeshContext) error {
-	for uri, resType := range rs.IndexByOrigin() {
-		conf := rules.Compute(uri, meshCtx.Resources)
-		if conf == nil {
-			conf = &outbound.ResourceRule{Conf: []interface{}{api.Conf{}}}
+func applyToRealResources(
+	rs []*core_xds.Resource,
+	rules outbound.ResourceRules,
+	reader kri.ResourceReader,
+) error {
+	for _, r := range rs {
+		switch r.ResourceOrigin.ResourceType {
+		case meshservice_api.MeshServiceType:
+		case meshexternalservice_api.MeshExternalServiceType:
+		case meshmultizoneservice_api.MeshMultiZoneServiceType:
+		default:
+			continue
 		}
 
-		for typ, resources := range resType {
-			switch typ {
-			case envoy_resource.ListenerType:
-				err := configureListeners(resources, conf.Conf[0].(api.Conf))
-				if err != nil {
-					return err
-				}
-			case envoy_resource.ClusterType:
-				err := configureClusters(resources, conf.Conf[0].(api.Conf))
-				if err != nil {
-					return err
-				}
-			}
+		var conf api.Conf
+		if rr := rules.Compute(*r.ResourceOrigin, reader); rr != nil {
+			conf = rr.Conf[0].(api.Conf)
 		}
-	}
-	return nil
-}
 
-func configureListeners(resources []*core_xds.Resource, conf api.Conf) error {
-	for _, resource := range resources {
-		configurer := plugin_xds.ListenerConfigurer{Conf: conf, Protocol: resource.Protocol}
-		err := configurer.ConfigureListener(resource.Resource.(*envoy_listener.Listener))
-		if err != nil {
-			return err
+		switch envoyResource := r.Resource.(type) {
+		case *envoy_listener.Listener:
+			configurer := plugin_xds.ListenerConfigurer{Conf: conf, Protocol: r.Protocol}
+			if err := configurer.ConfigureListener(envoyResource); err != nil {
+				return err
+			}
+
+			for _, fc := range envoyResource.FilterChains {
+				_ = listeners_v3.UpdateHTTPConnectionManager(fc, func(hcm *envoy_hcm.HttpConnectionManager) error {
+					if err := configureRoutes(hcm.GetRouteConfig(), rules, reader); err != nil {
+						return err
+					}
+					return nil
+				})
+			}
+
+		case *envoy_cluster.Cluster:
+			configurer := plugin_xds.ClusterConfigurerFromConf(conf, r.Protocol)
+			if err := configurer.Configure(envoyResource); err != nil {
+				return err
+			}
 		}
 	}
 	return nil
 }
 
-func configureClusters(resources []*core_xds.Resource, conf api.Conf) error {
-	for _, resource := range resources {
-		configurer := plugin_xds.ClusterConfigurerFromConf(conf, resource.Protocol)
-		err := configurer.Configure(resource.Resource.(*envoy_cluster.Cluster))
-		if err != nil {
-			return err
+func configureRoutes(
+	rc *envoy_route.RouteConfiguration,
+	rules outbound.ResourceRules,
+	reader kri.ResourceReader,
+) error {
+	for _, vh := range rc.VirtualHosts {
+		for _, route := range vh.Routes {
+			if !strings.HasPrefix(route.Name, "kri_") {
```

Comment:
Maybe it's not a question for this but does it work for gateway? I think for gateway we set name of the VH/Route to hash of the match.

Reply:
Yeah it doesn't work for gateway until https://github.com/kumahq/kuma/issues/13267 is done. I'm going to create an issue to take care of MeshTimeout and MeshHTTPRoute for Builtin Gateway

Reply:
https://github.com/kumahq/kuma/issues/13519